### PR TITLE
Adding subsection on 'Upgrading' guide about foreign key support

### PR DESCRIPTION
**Pull request by andreynering**
https://github.com/rails/rails/issues/18360

Adding this to help users that were using Foreigner on Rails < 4.2

### CODE REVIEW

File: guides/source/upgrading_ruby_on_rails.md [hunk 1]
```diff
@@ -276,6 +276,22 @@ class Notifier < ActionMailer::Base
...
 ```
 
+### Foreign Key support
+
+Rails 4.2 adds basic support for foreign keys. If the
+[Foreigner gem](https://github.com/matthuhiggins/foreigner) were been used to
+get support for foreign keys in the app, you can consider removing the gem
+from Gemfile on upgrading to 4.2 and using the default support.
+
+But remember that some optional and adapter specific options can not work as
```
Comment (jvanbaarsen) [inline guides/source/upgrading_ruby_on_rails.md]: But remember that some optional and adapter specific options **might** not work as expected

File: guides/source/upgrading_ruby_on_rails.md [hunk 2]
```diff
@@ -276,6 +276,22 @@ class Notifier < ActionMailer::Base
...
+expected, since the Rails has just basic support. The namming convention for
```
Comment (jvanbaarsen) [inline guides/source/upgrading_ruby_on_rails.md]: since rails has just basic support.

Comment (jvanbaarsen) [inline guides/source/upgrading_ruby_on_rails.md]: The naming convention

File: guides/source/upgrading_ruby_on_rails.md [hunk 3]
```diff
@@ -276,6 +276,22 @@ class Notifier < ActionMailer::Base
...
+generating the constraint names is also different, but this shouldn't cause
+problems.
```
Comment (jvanbaarsen) [inline guides/source/upgrading_ruby_on_rails.md]: Maybe explain in short what has changed?

Comment (andreynering) [inline guides/source/upgrading_ruby_on_rails.md]: @jvanbaarsen Thank you. Updated.

**Comment by jvanbaarsen** [comment]
@senny Looks good to me!

**Comment by senny** [comment]
@jvanbaarsen After a first glance there are some changes I'd like to propose. I'm off for today and will review this tomorrow.

File: guides/source/upgrading_ruby_on_rails.md [hunk 4]
```diff
@@ -276,6 +276,22 @@ class Notifier < ActionMailer::Base
...
+
+Rails 4.2 adds basic support for foreign keys. If the
+[Foreigner gem](https://github.com/matthuhiggins/foreigner) were been used to
+get support for foreign keys in the app, you can consider removing the gem
+from Gemfile on upgrading to 4.2 and using the default support.
+
+But remember that some optional and adapter specific options might not work as
+expected, since Rails has just basic support. The naming convention for
+generating the constraint names is also different (the names are randomly
+generated), but this shouldn't cause problems.
```
Comment (senny) [inline guides/source/upgrading_ruby_on_rails.md]: the names are no longer randomly generated. They are still not human readable but are consistent across migration runs.

File: guides/source/upgrading_ruby_on_rails.md [hunk 5]
```diff
@@ -276,6 +276,22 @@ class Notifier < ActionMailer::Base
 end
 ```
 
+### Foreign Key support
+
+Rails 4.2 adds basic support for foreign keys. If the
+[Foreigner gem](https://github.com/matthuhiggins/foreigner) were been used to
+get support for foreign keys in the app, you can consider removing the gem
+from Gemfile on upgrading to 4.2 and using the default support.
```
Comment (senny) [inline guides/source/upgrading_ruby_on_rails.md]: What about:

> The migration DSL has been expanded to support foreign key definitions. If you've been using the Foreigner gem, you might want to consider removing it. Note that the foreign key support of Rails is a subset of Foreigner.

**Comment by senny** [comment]
summoning @zzak to hopefully help with appropriate wording :hamster:

File: guides/source/upgrading_ruby_on_rails.md [hunk 6]
```diff
@@ -276,6 +276,22 @@ class Notifier < ActionMailer::Base
...
+
+But remember that some optional and adapter specific options might not work as
+expected, since Rails has just basic support. The naming convention for
+generating the constraint names is also different (the names are randomly
+generated), but this shouldn't cause problems.
+
+After removing Foreigner from Gemfile and running `bundle install`, it's
+important to run `rake db:schema:dump` and check if all foreign keys were
+dumped in `db/schema.rb`.
```
Comment (senny) [inline guides/source/upgrading_ruby_on_rails.md]: would be nice to group the necessary steps to replace the gem in one place. Something along the lines of:

1. remove `gem "foreigner"` from the Gemfile
1. run `bundle install`
1. run `bin/rake db:schema:dump`
1. make sure that `db/schema.rb` contains every foreign key definition with the necessary options.

Comment (andreynering) [inline guides/source/upgrading_ruby_on_rails.md]: @senny updated

**Comment by senny** [comment]
merged with some minor modifications.

@andreynering thank you.

**Pull request merged**